# Random Forest Regressor

In [20]:
%matplotlib inline
import numpy as np
import pandas as pd
from pandas import Series, DataFrame
import pickle
from sklearn.metrics import mean_squared_error as mse
from sklearn.model_selection import cross_val_score,KFold,GridSearchCV
from sklearn.ensemble import RandomForestRegressor
import warnings


warnings.filterwarnings("ignore")

# environment settings
data_path = 'Data/'

# Deserialize previously saved data from "preprocessing"
with open(data_path+'train_pp.obj', 'rb') as train_pp, \
open(data_path+'test_pp.obj','rb') as test_pp:
    train_df = pickle.load(train_pp)
    test_df = pickle.load(test_pp)
train_df["SalePrice"] = np.log1p(train_df["SalePrice"])

In [21]:
X = train_df.loc[:,'Id':'SaleCondition_Partial']
y = train_df['SalePrice']
print("Shape of training set {}.\nShape of test set {}".format(X.shape,y.shape))

Shape of training set (1460, 303).
Shape of test set (1460,)


In [28]:
# # Set the parameters by cross-validation
# tuned_parameters = {'n_estimators': [10,100,500], 
#                     'max_depth': [4, 6, 8], 
#                     'max_features':[50,100,150,200,250]}

# grid_rf = GridSearchCV(RandomForestRegressor(), 
#                        tuned_parameters, 
#                        cv=5, 
#                        n_jobs=-1, verbose=1)
# grid_rf.fit(X,y)
# print(grid_rf.best_score_)
# print(grid_rf.best_params_)

Fitting 5 folds for each of 45 candidates, totalling 225 fits


[Parallel(n_jobs=-1)]: Done  81 tasks      | elapsed:   38.6s
[Parallel(n_jobs=-1)]: Done 225 out of 225 | elapsed:  3.2min finished


0.8768506623465587
{'max_depth': 8, 'max_features': 100, 'n_estimators': 500}


## Found by grid search
r2 : 0.8768506623465587 <br>
{'max_depth': 8, 'max_features': 100, 'n_estimators': 500}

In [31]:
rf = RandomForestRegressor(n_estimators=500,
                           criterion="mse" ,
                           max_features=120,
                           max_depth=8,
                           random_state=1)
kf = KFold(10, shuffle=True, random_state=42).get_n_splits(X)
rmse_cv_rf= np.sqrt(-cross_val_score(rf, X, y, scoring="neg_mean_squared_error", cv = kf))
print("The 10-fold crossvalidation RMSE of Random Forest is {:.5f} +/- {:.3f}".format(rmse_cv_rf.mean(),
                                                                                          rmse_cv_rf.std()))

The 10-fold crossvalidation RMSE of Random Forest is 0.13853 +/- 0.016
